src/prefect/concurrency/v1/services.py

```diff
@@ -0,0 +1,103 @@
+import asyncio
+import concurrent.futures
+from contextlib import asynccontextmanager
+from typing import (
+    TYPE_CHECKING,
+    AsyncGenerator,
+    FrozenSet,
+    Optional,
+    Tuple,
+)
+from uuid import UUID
+
+import httpx
+from starlette import status
+
+from prefect._internal.concurrency import logger
+from prefect._internal.concurrency.services import QueueService
+from prefect.client.orchestration import get_client
+from prefect.utilities.timeout import timeout_async
+
+if TYPE_CHECKING:
+    from prefect.client.orchestration import PrefectClient
+
+
+class ConcurrencySlotAcquisitionService(QueueService):
+    def __init__(self, concurrency_limit_names: FrozenSet[str]):
+        super().__init__(concurrency_limit_names)
+        self._client: "PrefectClient"
+        self.concurrency_limit_names = sorted(list(concurrency_limit_names))
+
+    @asynccontextmanager
+    async def _lifespan(self) -> AsyncGenerator[None, None]:
+        async with get_client() as client:
+            self._client = client
+            yield
+
+    async def _handle(
+        self,
+        item: Tuple[
+            int,
+            str,
+            Optional[float],
+            concurrent.futures.Future,
+            Optional[bool],
+            Optional[str],
+        ],
+    ) -> None:
+        occupy, mode, timeout_seconds, future, create_if_missing, holder = item
+        try:
+            response = await self.acquire_slots(
+                occupy, mode, timeout_seconds, create_if_missing, holder
+            )
+        except Exception as exc:
+            # If the request to the increment endpoint fails in a non-standard
+            # way, we need to set the future's result so that the caller can
+            # handle the exception and then re-raise.
+            future.set_result(exc)
+            raise exc
+        else:
+            future.set_result(response)
+
+    async def acquire_slots(
+        self,
+        slots: int,
+        mode: str,
+        timeout_seconds: Optional[float] = None,
+        create_if_missing: Optional[bool] = False,
+        holder: Optional[str] = None,
+    ) -> httpx.Response:
+        with timeout_async(seconds=timeout_seconds):
+            while True:
+                try:
+                    response = await self._client.increment_concurrency_slots(
+                        names=self.concurrency_limit_names,
+                        slots=slots,
+                        mode=mode,
+                        create_if_missing=create_if_missing,
+                        holder=holder,
+                    )
+                except Exception as exc:
+                    if (
+                        isinstance(exc, httpx.HTTPStatusError)
+                        and exc.response.status_code == status.HTTP_423_LOCKED
+                    ):
+                        retry_after = float(exc.response.headers["Retry-After"])
```

Comment:
One small note, there's a case where we actually need to just stop the task entirely, so in that case you'll get a 423 _without a Retry-After_.  That happens when the limit exists and it actually set to `0`.  I think that's a way people can disable tasks without redeploying, kinda like a big red panic button.

In that case we probably need to raise something that will tell the user why?